### PR TITLE
US015 - Edit a role

### PR DESCRIPTION
**Pull request by RobertCrabtreeKainos**
added endpoint for edit role.

### CODE REVIEW

File: .github/workflows/integration-test.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+      - name: Create db.properties
+        run: |
+          echo "host=${{ secrets.DB_HOST }}" > db.properties && 
+          echo "name=${{ secrets.DB_NAME }}" >> db.properties && 
+          echo "user=${{ secrets.DB_USER }}" >> db.properties &&
+          echo "password=${{ secrets.DB_PASSWORD }}" >> db.properties
+
+      - name: Run integration tests
+        run: mvn -B integration-test --file pom.xml
```
Comment (shaunganley) [inline .github/workflows/integration-test.yml]: if you switch to using ENV VARs you wouldn't need to do this

File: src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        try {
+            return jobRoleDao.getJobRoles();
+        } catch (SQLException e) {
+            throw new FailedJobRolesOperationException("Failed to get job roles from database", e);
+        }
+    }
+
+    public int updateJobRole(Short id, UpdateJobRoleRequest jobRoleRequest) throws FailedJobRolesOperationException, FailedToUpdateJobRoleException {
+        try{
+             if(jobRoleDao.updateJobRole(id, jobRoleRequest) == id){
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/JobRoleService.java]: shouldn't need this logic. I'm assuming you're returning the ID from the DAO just to check this? if an exception isn't thrown you can assume the call was successful

File: src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 3]
```diff
@@ -25,4 +29,16 @@ public List<JobRoleResponse> getJobRoles() throws FailedJobRolesOperationExcepti
         }
     }
 
+    public int updateJobRole(Short id, UpdateJobRoleRequest jobRoleRequest) throws FailedJobRolesOperationException, FailedToUpdateJobRoleException {
+        try{
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/JobRoleService.java]: need validation before calling DAO

**Review by shaunganley — CHANGES_REQUESTED** [review]
add validation with integration test and unit tests to cover

File: .github/workflows/integration-test.yml (hunk 1 above)
Comment (RobertCrabtreeKainos) [reply .github/workflows/integration-test.yml]: its outdated

File: src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 4]
```diff
@@ -25,4 +29,11 @@ public List<JobRoleResponse> getJobRoles() throws FailedJobRolesOperationExcepti
         }
     }
 
+    public int updateJobRole(Short id, UpdateJobRoleRequest jobRoleRequest) throws FailedJobRolesOperationException, FailedToUpdateJobRoleException {
+        try {
+            return jobRoleDao.updateJobRole(id, jobRoleRequest);
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/api/JobRoleService.java]: You should probably make use of your validator.

File: src/main/java/org/kainos/ea/db/JobRoleDao.java [hunk 5]
```diff
@@ -53,4 +53,33 @@ public List<JobRoleResponse> getJobRoles() throws SQLException {
...
+        preparedStatement.setInt(3, jobRoleRequest.getBand().getBandID());
+        preparedStatement.setInt(4, jobRoleRequest.getCapability().getCapabilityID());
+        preparedStatement.setString(5, jobRoleRequest.getResponsibilities());
+        preparedStatement.setString(6, jobRoleRequest.getSharePoint());
+        preparedStatement.setInt(7, id);
+
+        int rowsUpdated = preparedStatement.executeUpdate();
+        if(rowsUpdated > 0)
+        {
+            System.err.println("hi ho ");
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/JobRoleDao.java]: it's not in the code you go

File: src/main/java/org/kainos/ea/resources/BandController.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    @ApiOperation(value = "Returns all Capabilities", authorizations = @Authorization(value = HttpHeaders.AUTHORIZATION), response = JobBandResponse.class)
+    public Response getJobRoles() {
+        try {
+            return Response.ok(bandService.getBands()).build();
+        } catch (FailedToGetBandsException e) {
+            e.printStackTrace();
+
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/resources/BandController.java]: For a SQL Exception, you should probably return a 500 error

File: src/main/java/org/kainos/ea/resources/JobRoleController.java [hunk 7]
```diff
@@ -39,4 +46,20 @@ public Response getJobRoles() {
         }
     }
 
+    @PUT
+    @Path("/job-roles/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    //@RolesAllowed("Admin")
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/resources/JobRoleController.java]: Commented out annotation on this line and line 54

File: src/main/java/org/kainos/ea/validator/UpdateJobRoleValidator.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        if (jobRole.getBand() == null || jobRole.getBand().getBandID() > 32767) {
+            return false;
+        }
+        if (jobRole.getCapability() == null || jobRole.getCapability().getCapabilityID() > 32767) {
+            return false;
+        }
+        if (jobRole.getResponsibilities() == null) {
+            return false;
+        }
+        if (jobRole.getSharePoint() == null || jobRole.getSharePoint().length() > 255) {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/validator/UpdateJobRoleValidator.java]: You probably want to validate this is a URL

File: src/main/java/org/kainos/ea/validator/UpdateJobRoleValidator.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        if (jobRole == null) {
+            return false;
+        }
+        if (jobRole.getJobRoleName() == null || jobRole.getJobRoleName().length() > 64) {
+            return false;
+        }
+        if (jobRole.getJobSpecSummary() == null) {
+            return false;
+        }
+        if (jobRole.getBand() == null || jobRole.getBand().getBandID() > 32767) {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/validator/UpdateJobRoleValidator.java]: For band and capability you probably want to check the value exists in the database as well as checking it's a number, so you return a 400 not a 500 back to the end user

File: src/test/java/org/kainos/ea/dao/BandDaoTest.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+
+public class BandDaoTest {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/dao/BandDaoTest.java:23]: Missing test case:
-> On DB error SQL Exception is thrown

File: src/test/java/org/kainos/ea/dao/CapabilityDaoTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+
+public class CapabilityDaoTest {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/dao/CapabilityDaoTest.java:22]: Missing test case:
-> On DB error SQL Exception is thrown

File: src/test/java/org/kainos/ea/dao/JobRoleDaoTest.java [hunk 12]
```diff
@@ -94,4 +97,36 @@ public void getJobRoles_shouldReturnEmptyJobRoles_whenThereAreNoJobRoles() throw
...
+        Connection connectionMock = mock(Connection.class);
+        Mockito.when(databaseConnector.getConnection()).thenReturn(connectionMock);
+        PreparedStatement mockPreparedStatement = mock(PreparedStatement.class);
+        Mockito.when(connectionMock.prepareStatement(any(String.class))).thenReturn(mockPreparedStatement);
+        Mockito.when(mockPreparedStatement.executeUpdate()).thenReturn(1);
+        int id = jobRoleDao.updateJobRole(1, jobRoleRequest);
+        assertEquals(1, id, "Expected 1 row to be updated.");
+    }
+    @Test
+    public void updateJobRole_shouldReturnNegative1_whenCantUpdate() throws SQLException, FailedToUpdateJobRoleException {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/dao/JobRoleDaoTest.java]: Test Name Update -> It throws an exception not returning -1

File: src/test/java/org/kainos/ea/integration/CapabilityIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class CapabilityIntegrationTest {
+    static final DropwizardAppExtension<DropwizardWebServiceConfiguration> APP = new DropwizardAppExtension<>(
+            DropwizardWebServiceApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
+
+    @Test
+    void band_shouldReturn200() {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/integration/CapabilityIntegrationTest.java]: Fix test name

File: src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 14]
```diff
@@ -23,5 +28,31 @@ void jobRoles_shouldReturn200() {
...
     }
+    @Test
+    void updateJobRoles_shouldReturn200() {
+        JobBandResponse jobBandResponse = new JobBandResponse(1, "jobBand");
+        JobCapabilityResponse jobCapabilityResponse = new JobCapabilityResponse(1, "jobCapability");
+        UpdateJobRoleRequest jobRoleRequest = new UpdateJobRoleRequest("jobRoleName", "jobSpecSummary",
+                jobBandResponse, jobCapabilityResponse,
+                "jobResponsibility", "sharepointLink");
+
+        Response response = APP.client().target("http://localhost:8080" + "/api/job-roles/1")
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java]: Should use TARGET_DOMAIN env var

File: src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 15]
```diff
@@ -23,5 +28,31 @@ void jobRoles_shouldReturn200() {
...
+                .request(MediaType.APPLICATION_JSON)
+                .put(Entity.entity(jobRoleRequest, MediaType.APPLICATION_JSON));
+
+        Assertions.assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
+    }
+    @Test
+    void updateJobRoles_shouldReturn500_whenNullBody() {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java]: Test says should return 500 but it checks for a 400

File: src/test/java/org/kainos/ea/service/CapabilityServiceTests.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+import org.mockito.Mockito;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class CapabilityServiceTests {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/service/CapabilityServiceTests.java:17]: Fix test names in this file as they say bands

File: src/main/java/org/kainos/ea/validator/UpdateJobRoleValidator.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
+package org.kainos.ea.validator;
+
+import org.kainos.ea.cli.UpdateJobRoleRequest;
+
+import java.util.regex.Pattern;
+
+public class UpdateJobRoleValidator {
+    String urlRegex = "https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)";
+
+    public boolean validate(UpdateJobRoleRequest jobRole)
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/validator/UpdateJobRoleValidator.java]: The more information we give back to the API client, the easier for the consuming application to respond. I suggest tweaking this validation so the user receives more detailed validation messages.

1. Change this method to void but throw a Validation Exception
2. When validation fails throw the Validation Exception with the reason why the validation failed (the if condition in English)
3. Update the service class to throw the Validation Exception and the if checking if validation was successful.

Hopefully that makes sense

File: src/main/java/org/kainos/ea/db/JobRoleDao.java [hunk 18]
```diff
@@ -53,4 +53,33 @@ public List<JobRoleResponse> getJobRoles() throws SQLException {
         return jobRoles;
     }
 
+    public int updateJobRole(int id, UpdateJobRoleRequest jobRoleRequest) throws SQLException, FailedToUpdateJobRoleException {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/JobRoleDao.java:77]: An invalid foreign key will cause the application to return a 500 error when it should return a 400, as that is user input. Is there a way we can make sure the right responses are sent back to the consumer

File: src/test/java/org/kainos/ea/dao/BandDaoTest.java (hunk 10 above)
Comment (TomBillingtonUK) [reply src/test/java/org/kainos/ea/dao/BandDaoTest.java:23]: > Missing test case: -> On DB error SQL Exception is thrown

Comment not addressed

File: src/test/java/org/kainos/ea/dao/CapabilityDaoTest.java (hunk 11 above)
Comment (TomBillingtonUK) [reply src/test/java/org/kainos/ea/dao/CapabilityDaoTest.java:22]: > Missing test case: -> On DB error SQL Exception is thrown

Comment not addressed

File: src/test/java/org/kainos/ea/service/CapabilityServiceTests.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+    CapabilityDao capabilityDao = Mockito.mock(CapabilityDao.class);
+    CapabilityService capabilityService;
+
+    @Test
+    public void constructor_whenNullBCapabilityDao_shouldThrowNullPointerException(){
+        assertThrows(NullPointerException.class, () -> new CapabilityService(null));
+    }
+
+    @Test
+    public void getBands_shouldReturnCapabilities_whenSuccess() throws SQLException, FailedToGetCapabilitiesException {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/service/CapabilityServiceTests.java]: Test name:
`getCapabilities_shouldReturnCapabilities_whenSuccess()`

File: src/test/java/org/kainos/ea/service/CapabilityServiceTests.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        List<JobCapabilityResponse> response = new ArrayList<>();
+        response.add(new JobCapabilityResponse(1, "capability"));
+
+        Mockito.when(capabilityDao.getCapabilities()).thenReturn(response);
+        capabilityService = new CapabilityService(capabilityDao);
+        assertEquals(response, capabilityService.getCapabilities());
+    }
+
+    @Test
+    public void getBands_shouldThrowFailedToGetCapabilityException_whenSQLException() throws SQLException {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/service/CapabilityServiceTests.java]: Test name:
`getCapabilities_shouldThrowFailedToGetCapabilityException_whenSQLException`

File: src/test/java/validator/UpdateJobRoleValidatorTests.java [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+        assertFalse(validator.validate(validJobRole));
+    }
+
+    @Test
+    public void testValidate_LongSharePoint_ReturnsFalse() {
+        validJobRole.setSharePoint("ThisIsALongSharePointLinkThatExceedsTheMaximumAllowedLength ThisIsALongSharePointLinkThatExceedsTheMaximumAllowedLength ThisIsALongSharePointLinkThatExceedsTheMaximumAllowedLength ThisIsALongSharePointLinkThatExceedsTheMaximumAllowedLength ThisIsALongSharePointLinkThatExceedsTheMaximumAllowedLength ThisIsALongSharePointLinkThatExceedsTheMaximumAllowedLength");
+        assertFalse(validator.validate(validJobRole));
+    }
+
+    @ParameterizedTest
```
Comment (TomBillingtonUK) [inline src/test/java/validator/UpdateJobRoleValidatorTests.java:76]: 👍  Nice use of this to avoid duplicating the test

**Review by TomBillingtonUK — APPROVED** [review]

**Review by LMBishop — APPROVED** [review]

**Pull request merged**